File: 101-200/141-150/144-BSTPreorderTraversal/BSTPreorderTraversal.cpp
#include <queue>
#include <vector>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

using namespace std;

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> nums;
        preorderTraversalHelper(root, nums);
        return nums;
    }

    void preorderTraversalHelper(TreeNode* node, vector<int> &nums) {
        if (!node)
            return;

        nums.push_back(node->val);
        preorderTraversalHelper(node->left, nums);
        preorderTraversalHelper(node->right, nums);
    }
};

File: 101-200/141-150/144-BSTPreorderTraversal/BSTPreorderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """

        result = []
        self.helper(root, result)
        return result

    def helper(self, node, result):
        if node:
            result.append(node.val)
            self.helper(node.left, result)
            self.helper(node.right, result)
